StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c

```diff
@@ -0,0 +1,862 @@
+    MmIplBuildFvHob (FoundationHobList + UsedSize, MmFvBase, MmFvSize);
+  }
+
+  UsedSize += sizeof (EFI_HOB_FIRMWARE_VOLUME);
+
+  //
+  // Build ACPI S3 Enable HOB
+  //
+  HobLength = (UINT16)((sizeof (EFI_HOB_GUID_TYPE) + sizeof (ACPI_S3_ENABLE) + 0x7) & (~0x7));
+  if (*FoundationHobSize >= UsedSize + HobLength) {
+    MmIplBuildAcpiS3EnableHob (FoundationHobList + UsedSize, HobLength);
+  }
+
+  UsedSize += HobLength;
+
+  //
+  // Build SMM CPU sync configuration HOB
+  //
+  HobLength = (UINT16)((sizeof (EFI_HOB_GUID_TYPE) + sizeof (SMM_CPU_SYNC_CONFIG) + 0x7) & (~0x7));
+  if (*FoundationHobSize >= UsedSize + HobLength) {
+    MmIplBuildSmmCpuSyncConfigHob (FoundationHobList + UsedSize, HobLength);
+  }
+
+  UsedSize += HobLength;
+
+  //
+  // Build CPU SMM base HOB in MM HOB list
+  //
+  RemainingSize = GetRemainingHobSize (*FoundationHobSize, UsedSize);
+  MmIplCopyGuidHob (FoundationHobList + UsedSize, &RemainingSize, &gSmmBaseHobGuid, TRUE);
+  UsedSize += RemainingSize;
+
+  //
+  // Build SMRAM memory Hob in MM HOB list
+  //
+  RemainingSize = GetRemainingHobSize (*FoundationHobSize, UsedSize);
+  MmIplCopyGuidHob (FoundationHobList + UsedSize, &RemainingSize, &gEfiSmmSmramMemoryGuid, FALSE);
+  UsedSize += RemainingSize;
+
+  //
+  // Build Mp Information2 Hob in MM HOB list
+  //
+  RemainingSize = GetRemainingHobSize (*FoundationHobSize, UsedSize);
+  MmIplCopyGuidHob (FoundationHobList + UsedSize, &RemainingSize, &gMpInformation2HobGuid, TRUE);
+  UsedSize += RemainingSize;
+
+  //
+  // Build ACPI variable HOB
+  //
+  RemainingSize = GetRemainingHobSize (*FoundationHobSize, UsedSize);
+  MmIplCopyGuidHob (FoundationHobList + UsedSize, &RemainingSize, &gEfiAcpiVariableGuid, FALSE);
+  UsedSize += RemainingSize;
+
+  if (FeaturePcdGet (PcdCpuSmmProfileEnable)) {
+    //
+    // Build memory allocation and resource HOB for smm profile data
+    //
+    HobLength = (UINT16)((sizeof (EFI_HOB_MEMORY_ALLOCATION) + sizeof (EFI_HOB_RESOURCE_DESCRIPTOR) + 0x7) & (~0x7));
+    if (*FoundationHobSize >= UsedSize + HobLength) {
+      SmmProfileDataHob = MmIplBuildSmmProfileHobs (FoundationHobList + UsedSize, &HobLength);
+    }
+
+    UsedSize += HobLength;
+  }
+
+  RemainingSize = GetRemainingHobSize (*FoundationHobSize, UsedSize);
+  if (FixedPcdGetBool (PcdCpuSmmRestrictedMemoryAccess)) {
+    //
+    // Only unblocked memory regions are accessible
+    //
+    MmIplBuildResourceHobForUnblockedRegion (FoundationHobList + UsedSize, &RemainingSize);
+  } else {
+    //
+    // All system memory (DRAM) is accessible.
+    // When SMM Profile is enabled:
+    //   * Access to regions reported from MmPlatformHobProducerLib do not require logging.
+    //   * Access to other system memory requires logging.
+    //
+    MmIplBuildResourceHobForAllSystemMemory (FoundationHobList + UsedSize, &RemainingSize, PlatformHobList, PlatformHobSize, Block, SmmProfileDataHob);

```

Comment:
1. Pls rename it to MmramRange or something else related with Mmram. Block looks a little confused
2. The Block you retrived from MmIplAllocateMmramPage() only contains the largest unallocated MMRAM region instead of all the ranges in gEfiSmmSmramMemoryGuid HOB